src/openssl/ssl/hostname_validation.cr

```diff
@@ -0,0 +1,128 @@
+require "socket"
+require "openssl"
+
+# :nodoc:
+module OpenSSL::SSL::HostnameValidation
+  enum Result
+    Error
+    MalformedCertificate
+    MatchFound
+    MatchNotFound
+    NoSANPresent
+  end
+
+  # Matches hostname against Subject Alternate Name (SAN) or Common Name (CN)
+  # entries of certificate if no SAN entries could be found in the certificate.
+  def self.validate_hostname(hostname : String, server_cert : LibCrypto::X509)
+    return Result::Error if server_cert.null?
+    result = matches_subject_alternative_name(hostname, server_cert)
+    result = matches_common_name(hostname, server_cert) if result.no_san_present?
+    result
+  end
+
+  # Matches hostname against Subject Alternate Name (SAN) entries of certificate.
+  #
+  # Adapted from https://wiki.openssl.org/index.php/Hostname_validation
+  def self.matches_subject_alternative_name(hostname, server_cert : LibCrypto::X509)
+    san_names = LibCrypto.x509_get_ext_d2i(server_cert, LibCrypto::NID_subject_alt_name, nil, nil)
+    return Result::NoSANPresent if san_names.null?
+
+    LibCrypto.sk_num(san_names).times do |i|
+      current_name = LibCrypto.sk_value(san_names, i).as(LibCrypto::GENERAL_NAME*).value
+      next unless current_name.type == LibCrypto::GEN_DNS
+
+      dns_name = LibCrypto.asn1_string_data(current_name.dns_name)
+      dns_name_len = LibCrypto.asn1_string_length(current_name.dns_name)
+      return Result::MalformedCertificate if dns_name_len != LibC.strlen(dns_name)
+
+      pattern = String.new(dns_name, dns_name_len)
+      return Result::MatchFound if matches_hostname?(pattern, hostname)
```

Comment:
Yes, maybe OpenSSL did it already when it generated the `GENERAL_NAME` struct, or this is what the `ASN1_STRING` methods do.